### PR TITLE
feat: kotlin install waits on java install

### PR DESCRIPTION
**Pull request by dgschwindturo**
This implementation does the following:

- First and foremost, adds support for Kotlin via sdkman to this action
- Introduces abstract base class `SdkmanTool` that extends `Tool` that both `Java` and `Kotlin` extend.
- A Tool does not inheritly depend on another Tool (even though we definitely have such a case with Kotlin depending upon Java)
- Follows implementation approach suggested by @shakefu where the kotliln setup itself waits for the java install to have had a chance to succeed, before itself proceeding.

Adding Alfonso and Victor to review these heavy Javascript changes per @shakefu suggestion.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -17,26 +18,15 @@ export default async function run() {
...
-        for (let tool of Tool.all()) {
-            try {
-                // For some reason this catch call isn't working the way I expect,
-                // but I can't figure it out, so we double down with try/catch
-                await tool.setup(core.getInput(tool.name)).catch(errHandler)
-            } catch (err) {
-                errHandler(err)
-            }
-        }
+        await setupToolsSequentially(errHandler)
```
Comment (llosimura) [inline index.js:29]: What about handling the errors internally and return them?

File: java.js [hunk 2]
```diff
@@ -68,53 +68,49 @@ export default class Java extends SdkmanTool {
...
+     * presented to the action is present, that version is honored rather than the version
+     * presented in the .sdkmanrc file that can be optionally present in the checked out repo itself.
+     * Second value returned indicates whether or not the version returned has overridden
+     * the version from the .sdkmanrc file.
+     * @param {string} actionDesiredVersion - This is the desired version of java as presented directly to the
+     * action.
+     * @returns {[string, boolean]} - The overall desired version of java that has been found, and the boolean
+     * indicates whether or not that version is overriding the value found in the .sdkmanrc file.
+     */
     getJavaVersion(actionDesiredVersion) {
```
Comment (llosimura) [inline java.js:87]: It seems that there are some duplication between `getJavaVersion` and `getKotlinVersion` could we just override the `getVersion` and put it on the SdkmanTool?

**Review by llosimura — COMMENTED** [review]
Done with my first pass


I have an open questions about the dependency between Kotlin & Java. Per the PR description
> A Tool does not inheritly depend on another Tool (even though we definitely have such a case with Kotlin depending upon Java)

But we do indeed wait for Java to install Kotlin
1. What is the reason behind this?
2. The wait for will fail after 5 minutes. Since it is in fact a dependency, doesn't it make more sense to treat is a dependency and model it as such?

File: java.js (hunk 2 above)
Comment (dgschwindturo) [reply java.js:87]: Good point Alfonso, there is a good amount of duplication there. I will go about pushing a common method up to the superclass.

File: README.md [hunk 3]
```diff
@@ -150,6 +153,7 @@ jobs:
               with:
                   go: 1.17.6
                   java: 17.0.2-tem
+                  kotlin: 1.6.21
```
[the remark this answers is not in the document]
Comment (dgschwindturo) [reply README.md:156]: Good point, can tweak the docs to reflect 1.8.0

**Comment by dgschwindturo** [comment]
> Done with my first pass
> 
> I have an open questions about the dependency between Kotlin & Java. Per the PR description
> 
> > A Tool does not inheritly depend on another Tool (even though we definitely have such a case with Kotlin depending upon Java)
> 
> But we do indeed wait for Java to install Kotlin
> 
> 1. What is the reason behind this?
> 2. The wait for will fail after 5 minutes. Since it is in fact a dependency, doesn't it make more sense to treat is a dependency and model it as such?

**Comment by dgschwindturo** [comment]
> 



> Done with my first pass
> 
> I have an open questions about the dependency between Kotlin & Java. Per the PR description
> 
> > A Tool does not inheritly depend on another Tool (even though we definitely have such a case with Kotlin depending upon Java)
> 
> But we do indeed wait for Java to install Kotlin
> 
> 1. What is the reason behind this? [DG] Kotlin will fail to be installed if there is not a valid Java installation in place at the time Kotlin is attempting to be installed.
> 2. The wait for will fail after 5 minutes. Since it is in fact a dependency, doesn't it make more sense to treat is a dependency and model it as such? [DG] That was certainly considered, but the tricky part comes when attempting to install all of the tools in parallel. Thus, that is how we arrived at the implementation you see, rather than having the Tool class have a relationship to itself.

**Comment by tagoro9** [comment]
@dgschwindturo going to close this since it has been opened for a while. We can always reopen if still relevant

**Pull request closed without merging**